### DISCUSSION TITLE
Trouble adding dock widget / binaryninjaui.DockHandler already deleted

### DISCUSSION THREAD

**Discussion by lks128** [Vector35/binaryninja-api#4671] [API] [answered]
Dear community,

I created a simple widget which displays name of the highlighted function, however it looks that I am doing something wrong when I try to add it to the dock.

The widget class is defined in `my_widget.py` like this:

```python
from binaryninjaui import DockHandler, DockContextHandler, UIActionHandler

from PySide6 import QtCore
from PySide6.QtCore import Qt
from PySide6.QtWidgets import QApplication, QHBoxLayout, QVBoxLayout, QLabel, QWidget, QMainWindow


class HelloDockWidget(QWidget, DockContextHandler):
    def __init__(self, parent, name):
        QWidget.__init__(self, parent)
        DockContextHandler.__init__(self, self, name)

        self.actionHandler = UIActionHandler()
        self.actionHandler.setupActionHandler(self)

        self.label = QLabel(name, self)

        layout = QVBoxLayout()
        layout.addWidget(self.label)

        self.setLayout(layout)


    def notifyViewLocationChanged(self, view, view_location):
        action_context = view.actionContext()
        bv = action_context.binaryView
        token_state = action_context.token
        if token_state.valid:
            token = token_state.token
            fun = bv.get_function_at(token.value)
            if fun:
                self.label.setText(fun.name)

# second part follows here...
```

Now in the second part I tried two approaches. Using `DockHandler.getActiveDockHandler()` always gives me an exception.

```python
# ... file continues
dock_handler = DockHandler.getActiveDockHandler()
widget = HelloDockWidget(dock_handler.parent(), "Test")
dock_handler.addDockWidget(widget, Qt.AllDockWidgetAreas, Qt.Horizontal, True)
```

```
[Default] Traceback (most recent call last):
[Default]   File "C:\Users\Andris\AppData\Roaming\Binary Ninja\plugins\my_widget.py", line 37, in <module>
[Default]     dock_handler.addDockWidget(widget, Qt.AllDockWidgetAreas, Qt.Horizontal, True)
[Default] RuntimeError: Internal C++ object (binaryninjaui.DockHandler) already deleted.
[PythonPlugin] Python plugin 'my_widget' could not be loaded
```

Second approach is to use `QApplication.allWidgets()` and add `__class__` check, because sometimes first widget is a `QFrame` without dock handler.

```python
# ... file continues (working option)
window = [ w.window() for w in QApplication.allWidgets() if w.__class__ == QMainWindow ][0]
dock_handler = window.findChild(DockHandler, '__DockHandler')
widget = HelloDockWidget(dock_handler.parent(), "Test")
dock_handler.addDockWidget(widget, Qt.AllDockWidgetAreas, Qt.Horizontal, True)
```

Although this works, it feels somewhat wrong. I think I am missing some detail about where or how to trigger widget addition to the dock.

Search of existing plugins shows that they do indeed use `DockHandler.getActiveDockHandler()`, so I think it is not a bug, I am just trying to use it in a wrong way or at a wrong time.

Two relevant additional questions:
1. How do I make a dock disabled until a file is opened? Like on this screenshot from gaasedelen/lighthouse:

![image](https://github.com/Vector35/binaryninja-api/assets/120235/fcd6946d-4ee7-41ab-a4fd-dfa9ec92cd3e)

2. `notifyViewLocationChanged` is only fired when highlighted line is changed and not when selection is changed within the line. How do I hook to each selection change and not only on distinct lines? (Like cross references window)

Best regards,
Andris

**Comment by lks128** [comment]
**Update**

- After some digging I found out that the problem is with `dock_handler.parent()` and not the `DockHandler.getActiveDockHandler()` call itself. Using dock handler from that method and window from `QApplication.allWidgets()` works.

- The way to make it disabled until file is opened is to register the dock widget dynamically:

```python
# ... file continues (working dynamic option

def create_dockable_widget(name, parent, bv):
    return HelloDockWidget(parent, name)

dock_handler = DockHandler.getActiveDockHandler()
dock_handler.addDockWidget("Test", create_dockable_widget, QtCore.Qt.RightDockWidgetArea, QtCore.Qt.Horizontal, False)
```

Magically it also solves the issue with deleted parent. 🎉 

So the only left open point is about how to receive events on each token change:

> notifyViewLocationChanged is only fired when highlighted line is changed and not when selection is changed within the line. How do I hook to each selection change and not only on distinct lines? (Like cross references window)

**Reply by fuzyll** [reply, marked as answer]
Hey, so, `DockWidget`s are actually a pretty old construct of ours and they're not really actively maintained. We've basically moved to a pane system with a sidebar area instead. We'd recommend re-implementing this as a `SidebarWidget` instead (which also has a [`notifyViewLocationChanged`](https://api.binary.ninja/cpp/class_sidebar_widget.html#a8f9ebc7b30bf24453f7310380fdd1578), especially because `DockWidget`s will - at some point - be deprecated and removed.

It's entirely possible that this was a bug that is now fixed in the `SidebarWidget`'s implementation of that API. But, if it doesn't work there either, it's probably a limitation we haven't documented well enough.

If you do wind up reimplementing in terms of `SidebarWidget` and can let us know how it goes, that'd be great. 🙂
